packages/animations/lib/src/utils/composite_animation_widget.dart

```diff
@@ -0,0 +1,253 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:animations/animations.dart';
+import 'package:flutter/material.dart';
+
+/// Signature for a function that creates a widget that builds a
+/// enter transition.
+///
+/// Used by [CompositeAnimationWidget]
+typedef _EnterTransitionBuilder = Widget Function(
+    BuildContext, Animation<double>, Widget);
+
+/// Signature for a function that creates a widget that builds a
+/// exit transition.
+///
+/// Used by [CompositeAnimationWidget]
+typedef _ExitTransitionBuilder = Widget Function(
+    BuildContext, Animation<double>, Widget);
+
+/// A widget that animate it's child with different transitions when the
+/// [CompositeAnimationWidget.animation] value goes forward
+/// (built by [CompositeAnimationWidget.enterTransitionBuilder])
+/// and reverse
+/// (built by [CompositeAnimationWidget.exitTransitionBuilder]).
+///
+/// It is used in [FadeThroughTransition], [SharedAxisTransition],
+/// and [FadeScaleTransition] widgets.
+///
+/// This widget also preserve it's child state by building all of it's enter
+/// and exit transitions stacked and only changing animation values of
+/// both transitions instead of dynamically changing widget
+/// subtrees levels in [AnimatedBuilder] based on effective [AnimationStatus]
+/// causes the state object in that particular subtree level to get recreated.
+///
+/// The state preservation of the child will take effect if and only if you're
+/// not changing subtree levels in defined transition builders.
+class CompositeAnimationWidget extends StatefulWidget {
+  /// Creates a widget that animate it's child with different transitions when
+  /// the [animation] value goes forward
+  /// (built by [enterTransitionBuilder])
+  /// and reverse
+  /// (built by [exitTransitionBuilder]).
+  const CompositeAnimationWidget({
+    @required this.animation,
+    @required this.child,
+    @required this.enterTransitionBuilder,
+    @required this.exitTransitionBuilder,
+    this.flip = false,
+    this.syncVisibleStates = false,
+  })  : assert(child != null),
+        assert(animation != null),
+        assert(enterTransitionBuilder != null),
```

Comment:
The doc comment of the constructor should also document which arguments cannot be null.